### PR TITLE
Monitoring integration

### PR DESCRIPTION
**Pull request by enxebre**
This enables monitoring integration for cao.
- Controller runtime metrics are exposed for machine_autoscaler_controller and cluster_autoscaler_controller
- This enables any farther metric to be scraped by prometheus for either these controllers or any managed component i.e machine-provider/cluster-autoscaler

### CODE REVIEW

File: pkg/operator/config.go [hunk 1]
```diff
@@ -228,5 +235,15 @@ func ConfigFromEnvironment() *Config {
...
 	}
 
+	if metricsPort, ok := os.LookupEnv("METRICS_PORT"); ok {
+		v, err := strconv.ParseInt(metricsPort, 10, 32)
+		if err != nil {
+			v = DefaultMetricsPort
+			klog.Errorf("Error parsing METRICS_PORT environment variable: %v", err)
+		}
+
+		config.metricsPort = int(v)
```
Comment (frobware) [inline pkg/operator/config.go]: If the `ParseInt` fails then metricsPort is likely to (the zero value) 0. This will then listen on port 0 -- the ephemeral port number - and will appear to work as the OS will assign an unused port number. I don't think we want this. We should return an error if the conversion fails.

File: pkg/operator/config.go [hunk 2]
```diff
@@ -228,5 +235,15 @@ func ConfigFromEnvironment() *Config {
 		config.WebhooksCertDir = webhooksCertDir
 	}
 
+	if metricsPort, ok := os.LookupEnv("METRICS_PORT"); ok {
+		v, err := strconv.ParseInt(metricsPort, 10, 32)
+		if err != nil {
+			v = DefaultMetricsPort
+			klog.Errorf("Error parsing METRICS_PORT environment variable: %v", err)
```
Comment (frobware) [inline pkg/operator/config.go]: Log the value that was parsed, but failed:

`klog.Errorf("Error parsing %q METRICS_PORT environment variable: %v", os.LookupEnv("METRICS_PORT"), err)`

File: pkg/operator/config.go [hunk 3]
```diff
@@ -215,18 +223,26 @@ func ConfigFromEnvironment() *Config {
...
 
 	if webhooksCertDir, ok := os.LookupEnv("WEBHOOKS_CERT_DIR"); ok {
 		config.WebhooksCertDir = webhooksCertDir
 	}
 
-	return config
+	if metricsPort, ok := os.LookupEnv("METRICS_PORT"); ok {
+		v, err := strconv.Atoi(metricsPort)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing METRICS_PORT (%q) environment variable: %v", metricsPort, err)
```
Comment (frobware) [inline pkg/operator/config.go]: Let's raise the bar and add unit tests for the new behaviour, testing for both positive/negative results when setting `METRICS_PORT` and `WEBHOOKS_CERT_DIR`.

File: install/0000_90_cluster-autoscaler-operator_01_prometheus_rbac.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
Comment (bison) [inline install/0000_90_cluster-autoscaler-operator_01_prometheus_rbac.yaml]: Why are this file and the ServiceMonitor file named differently from the others? There is an existing file with RBAC rules.

File: pkg/operator/config_test.go [hunk 5]
```diff
@@ -12,3 +16,89 @@ func TestNewConfig(t *testing.T) {
...
+	for _, tc := range testCase {
+		for key, val := range tc.envVars {
+			os.Setenv(key, val)
+		}
+		got, err := ConfigFromEnvironment()
+		if (err != nil) != tc.expectedError {
+			t.Errorf("expected %v, got: %v", tc.expectedError, err)
+		}
+		if !reflect.DeepEqual(got, tc.expectedConfig) {
+			t.Errorf("Got: %v, expected: %v", got, tc.expectedConfig)
```
Comment (frobware) [inline pkg/operator/config_test.go]: Keep the args the same as the previous `t.Errorf()` for consistency (i.e., expected, got).

File: install/0000_90_cluster-autoscaler-operator_02_servicemonitor.yaml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  labels:
+    k8s-app: cluster-autoscaler-operator
+  name: cluster-autoscaler-operator
+  namespace: openshift-monitoring
```
Comment (bison) [inline install/0000_90_cluster-autoscaler-operator_02_servicemonitor.yaml]: This seems wrong. Shouldn't this go in our namespace?

File: pkg/controller/clusterautoscaler/clusterautoscaler_controller.go [hunk 7]
```diff
@@ -28,7 +28,7 @@ import (
 )
 
 const (
-	controllerName      = "cluster-autoscaler-controller"
+	controllerName      = "cluster_autoscaler_controller"
```
Comment (bison) [inline pkg/controller/clusterautoscaler/clusterautoscaler_controller.go]: I know why you're doing this, but I really wish it was fixed in controller-runtime...

File: pkg/operator/config.go [hunk 8]
```diff
@@ -122,6 +124,9 @@ type Config struct {
 	// WebhookCertDir is the directory containing TLS assets for the admission
 	// webhook server.
 	WebhooksCertDir string
+
+	// metricsPort is the port the metrics are exposed.
+	metricsPort int
```
Comment (bison) [inline pkg/operator/config.go]: Why would the metrics port be unexported unlike all the other fields?

Comment (enxebre) [reply pkg/operator/config.go]: any reason for them to be exported? should we instead unexport in a different PR?

Comment (bison) [reply pkg/operator/config.go]: They just seem like public fields to me. This is not something that's going to cross system boundaries. We don't have to be super protective of it, but if I want to pass it to another package within the project at some point, I should be able to read the fields.

Comment (enxebre) [reply pkg/operator/config.go]: ok sg

File: install/03_rbac.yaml [hunk 9]
```diff
@@ -221,3 +221,37 @@ subjects:
   - kind: ServiceAccount
     name: cluster-autoscaler
     namespace: openshift-machine-api
+
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: prometheus-k8s
```
Comment (bison) [inline install/03_rbac.yaml]: A bit concerned about naming of the Role and RoleBinding. We share a namespace with all the other cluster-api related components. Will they each create this when they have monitoring enabled? If so, they should probably have the operator name included here to create separate resources, even if the permissions are the same.

Comment (enxebre) [reply install/03_rbac.yaml]: great point
